### DISCUSSION TITLE
Creating custom Emmets / Snippets

### DISCUSSION THREAD

**Discussion by VishwaKumar** [microsoft/vscode-discussions#15] [Extension Development QnA] [answered]
Hi,

I have this specific use case where from my extension, I want users to be able to type `!myamltemplate` in a yaml file and be able to load in a designated yaml template for users to start editing. This is similar to the `!` Emmet that VSCode currently supports.
So is there a way to create my custom Emmets and if not what's the best way to implement this use case.

Edit1:
I see we have this doc on how to use [custom Emmets](https://code.visualstudio.com/docs/editor/emmet#_using-custom-emmet-snippets) but the example shows using a html one. 
- Can this be used for yaml files as well ?
- Can the yaml template be loaded from a file 
- I would like to also have diagnostics and validations on the user edited values in yaml file

Edit2:
I was able to achieve my use case with [Snippets](https://code.visualstudio.com/api/language-extensions/snippet-guide)

```
// package.json
"contributes": {
	"snippets": [
		{
			"language": "yaml",
			"path": "./snippets.json"
		}
	]
}
```

```
// snippets.json
{
	"yaml": {
		"prefix": "!mytemplate",
		"body": [
			"name: install and configure DB",
			"hosts: testServer",
			"vars:",
			"  oracle_db_port_value : 1521",
			"tasks:",
			"  -name: Install the Oracle DB",
			"     yum: <code to install the DB>",
			" ",
			"  -name: Ensure the installed service is enabled and running",
			"service:",
			"   name: <your service name>"
		],
		"description": "Creates a yaml definition from template"
	}
}
```

**Questions**
- Is this the best approach for this use case ?
- Where can I find samples for `.tmSnippets` as described herein [Using TextMate snippets](https://code.visualstudio.com/api/language-extensions/snippet-guide#using-textmate-snippets) 
Thank you!

**Comment by rzhao271** [comment, marked as answer]
Emmet snippets can really only be used for HTML and CSS. For other languages, it'll be best to use the snippets feature. Emmet HTML abbreviations also have their own syntax, such as `ul>li`.
I'm unable to find a sample for `.tmSnippets` but there is this Markdown file that explains the snippets syntax some more: https://github.com/microsoft/vscode/blob/main/src/vs/editor/contrib/snippet/browser/snippet.md
